### PR TITLE
Integrate XCM

### PR DESCRIPTION
**Pull request by drgora**
This PR is not on main, but it is targeting the feature branch for parachains!

Adds support for XCM to the relay chain and the test parachain; some basic configuration is added to allow for two-way teleports of ACME between the relay chain and the test parachain.

### CODE REVIEW

File: runtime/src/parachains.rs [hunk 1]
```diff
@@ -123,54 +124,56 @@ impl paras::Config for Runtime {
...
+    /// # WARNING
+    ///
+    /// This is not a good value for para-chains since the `Scheduler` already uses up to 80% block weight.
+    pub MessageQueueServiceWeight: Weight = Perbill::from_percent(20) * BlockWeights::get().max_block;
+    pub MessageQueueIdleServiceWeight: Weight = Perbill::from_percent(20) * BlockWeights::get().max_block;
+    pub const MessageQueueHeapSize: u32 = 32 * 1024;
+    pub const MessageQueueMaxStale: u32 = 96;
+}
 
 // /// Message processor to handle any messages that were enqueued into the `MessageQueue` pallet.
```
Comment (la10736) [inline runtime/src/parachains.rs:139]: please un comment this

File: runtime/src/parachains.rs [hunk 2]
```diff
@@ -123,54 +124,56 @@ impl paras::Config for Runtime {
...
+impl pallet_message_queue::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type Size = u32;
+    type HeapSize = MessageQueueHeapSize;
+    type MaxStale = MessageQueueMaxStale;
+    type ServiceWeight = MessageQueueServiceWeight;
+    type MessageProcessor = MessageProcessor;
+    //#[cfg(feature = "runtime-benchmarks")]
+    //type MessageProcessor =
+    //    pallet_message_queue::mock_helpers::NoopMessageProcessor<AggregateMessageOrigin>;
```
Comment (la10736) [inline runtime/src/parachains.rs:171]:
```suggestion
    #[cfg(not(feature = "runtime-benchmarks")))]
    type MessageProcessor = MessageProcessor;
    #[cfg(feature = "runtime-benchmarks")]
    type MessageProcessor =
        pallet_message_queue::mock_helpers::NoopMessageProcessor<AggregateMessageOrigin>;
```

**Review by la10736 — APPROVED** [review]
In general LGTM

File: paratest/runtime/src/lib.rs [hunk 3]
```diff
@@ -44,10 +44,10 @@ use frame_system::{
     EnsureRoot,
 };
 // use pallet_xcm::{EnsureXcm, IsVoiceOfBody};
```
[the remark this answers is not in the document]
Comment (drgora) [reply paratest/runtime/src/lib.rs:46]: Done

File: paratest/runtime/src/xcm_config.rs [hunk 4]
```diff
@@ -1,142 +1,158 @@
...
-	RelayChainAsNative, SiblingParachainAsNative, SiblingParachainConvertsVia,
-	SignedAccountId32AsNative, SignedToAccountId32, SovereignSignedViaLocation, TakeWeightCredit,
-	TrailingSetTopicAsId, UsingComponents, WithComputedOrigin, WithUniqueTopic,
+    AccountId32Aliases, AllowUnpaidExecutionFrom, EnsureXcmOrigin, FixedWeightBounds,
+    FrameTransactionalProcessor, FungibleAdapter, IsConcrete, MintLocation, NativeAsset,
+    ParentIsPreset, RelayChainAsNative, SiblingParachainAsNative, SignedAccountId32AsNative,
+    SignedToAccountId32, SovereignSignedViaLocation, UsingComponents, WithUniqueTopic,
 };
 use xcm_executor::XcmExecutor;
+//use xcm::opaque::v3::{InteriorLocation, Location};
```
[the remark this answers is not in the document]
Comment (drgora) [reply paratest/runtime/src/xcm_config.rs:30]: Done

File: paratest/runtime/src/xcm_config.rs [hunk 5]
```diff
@@ -1,142 +1,158 @@
...
-	pub const RelayNetwork: Option<NetworkId> = None;
-	pub RelayChainOrigin: RuntimeOrigin = cumulus_pallet_xcm::Origin::Relay.into();
-	pub UniversalLocation: InteriorMultiLocation = Parachain(ParachainInfo::parachain_id().into()).into();
+    pub const RelayLocation: Location = Location::parent();
+    pub const RelayNetwork: Option<NetworkId> = Some(NetworkId::Polkadot);
+    pub RelayChainOrigin: RuntimeOrigin = cumulus_pallet_xcm::Origin::Relay.into();
+    pub UniversalLocation: InteriorLocation = Parachain(ParachainInfo::parachain_id().into()).into();
+  pub CheckAccount: AccountId = XcmPallet::check_account();
+  pub LocalCheckAccount: (AccountId, MintLocation) = (CheckAccount::get(), MintLocation::Local);
 }
```
[the remark this answers is not in the document]
Comment (drgora) [reply paratest/runtime/src/xcm_config.rs:40]: Good catch, fixed it

File: paratest/runtime/src/xcm_config.rs [hunk 6]
```diff
@@ -1,142 +1,158 @@
...
 
-/// Type for specifying how a `MultiLocation` can be converted into an `AccountId`. This is used
```
[the remark this answers is not in the document]
Comment (drgora) [reply paratest/runtime/src/xcm_config.rs:32]: We implement the latest version (v4) of xcm, which renamed `MultiLocation` and MultiAsset` into `Location` and `Asset` respectvely, along with others.

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -352,6 +355,7 @@ impl pallet_balances::Config for Runtime {
 
 parameter_types! {
     pub FeeMultiplier: Multiplier = Multiplier::one();
+    pub storage TransactionByteFee: Balance = 10 * MILLICENTS;
```
[the remark this answers is not in the document]
Comment (drgora) [reply runtime/src/lib.rs:358]: Agreed and done

File: runtime/src/parachains.rs (hunk 2 above)
Comment (drgora) [reply runtime/src/parachains.rs:171]: Done

**Pull request merged**